### PR TITLE
Add support for estimating remaining render time in appleseed.studio

### PR DESCRIPTION
**Pull request by achalpandeyy**
Implements #2438 

Supersedes #2454

### CODE REVIEW

File: src/appleseed.cli/main.cpp [hunk 1]
```diff
@@ -29,12 +29,12 @@
 
 // appleseed.cli headers.
 #include "commandlinehandler.h"
-#include "progresstilecallback.h"
 #include "stdouttilecallback.h"
 
 // appleseed.shared headers.
 #include "application/application.h"
 #include "application/superlogger.h"
+#include "application/progresstilecallback.h"
```
Comment (dictoon) [inline src/appleseed.cli/main.cpp]: Order alphabetically.

File: src/appleseed.shared/application/progresstilecallback.cpp [hunk 2]
```diff
@@ -64,9 +64,8 @@ namespace
       : public TileCallbackBase
     {
       public:
-        explicit ProgressTileCallback(Logger& logger)
-          : m_logger(logger)
-          , m_rendered_pixels(0)
+        explicit ProgressTileCallback()
```
Comment (dictoon) [inline src/appleseed.shared/application/progresstilecallback.cpp]: It was correct before, appleseed.shared shouldn't depend on the renderer's logger (`RENDERER_LOG_*`) which as its name implies, is specific to the renderer. Revert these changes.

In appleseed.renderer and appleseed.studio, you can use `renderer::global_logger()` (declared in `renderer/global/globallogger.h`) to access the renderer's global logger.

File: src/appleseed.shared/application/progresstilecallback.h [hunk 3]
```diff
@@ -29,23 +29,23 @@
...
 
-// Forward declarations.
-namespace foundation    { class Logger; }
-
 namespace appleseed {
-namespace cli {
+namespace shared {
 
-class ProgressTileCallbackFactory
+class SHAREDDLL ProgressTileCallbackFactory
```
Comment (dictoon) [inline src/appleseed.shared/application/progresstilecallback.h]: If you expose this class, then it cannot use anything from the `std::` namespace. You'll need to provide a private implementation.

File: src/appleseed.studio/mainwindow/rendering/renderingmanager.h [hunk 4]
```diff
@@ -162,15 +161,16 @@ class RenderingManager
 
   private:
     StatusBar&                                  m_status_bar;
-    QtRendererController                        m_renderer_controller;
+    QtRendererController                        m_renderer_controller;
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/rendering/renderingmanager.h]: Revert.

File: src/appleseed.studio/mainwindow/rendering/renderingmanager.h [hunk 5]
```diff
@@ -162,15 +161,16 @@ class RenderingManager
...
+    QtRendererController                        m_renderer_controller; 
 
     renderer::Project*                          m_project;
     renderer::ParamArray                        m_params;
     foundation::SearchPaths                     m_resource_search_paths;
     RenderingMode                               m_rendering_mode;
     RenderTab*                                  m_render_tab;
 
-    std::unique_ptr<QtTileCallbackFactory>      m_tile_callback_factory;
+    std::unique_ptr<renderer::TileCallbackCollectionFactory>
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/rendering/renderingmanager.h]: You need to include the proper header.

File: src/appleseed/CMakeLists.txt [hunk 6]
```diff
@@ -1299,6 +1299,8 @@ set (renderer_kernel_rendering_sources
     renderer/kernel/rendering/tilecallbackbase.h
     renderer/kernel/rendering/timedrenderercontroller.cpp
     renderer/kernel/rendering/timedrenderercontroller.h
+    renderer/kernel/rendering/tilecallbackcollection.cpp
```
Comment (dictoon) [inline src/appleseed/CMakeLists.txt]: Order alphabetically.

File: src/appleseed/renderer/api/rendering.h [hunk 7]
```diff
@@ -46,3 +46,4 @@
 #include "renderer/kernel/rendering/progressive/progressiveframerenderer.h"
 #include "renderer/kernel/rendering/tilecallbackbase.h"
 #include "renderer/kernel/rendering/timedrenderercontroller.h"
+#include "renderer/kernel/rendering/tilecallbackcollection.h"
```
Comment (dictoon) [inline src/appleseed/renderer/api/rendering.h]: Order alphabetically.

File: src/appleseed/renderer/kernel/rendering/tilecallbackcollection.h [hunk 8]
```diff
@@ -0,0 +1,66 @@
+
+//
+// This source file is part of appleseed.
+// Visit https://appleseedhq.net/ for additional information and resources.
+//
+// This software is released under the MIT license.
+//
+// Copyright (c) 2010-2013 Francois Beaune, Jupiter Jazz Limited
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/tilecallbackcollection.h]: Replace these two lines by a single one:

    Copyright (c) 2019 YOUR NAME, The appleseedhq Organization

Replace `YOUR NAME` with your name.

Same change in the .cpp file.

File: src/appleseed/renderer/kernel/rendering/tilecallbackcollection.cpp [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+namespace
+{
+    class TileCallbackCollection
+      : public ITileCallback
+    {
+      public:
+        explicit TileCallbackCollection(list<ITileCallbackFactory*> factories)
+        {
+            for (each<list<ITileCallbackFactory*>> i = factories; i; ++i)
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/tilecallbackcollection.cpp]: Switch to range-based `for` loops in new code, and get rid of `foundation/utility/foreach.h`.

File: src/appleseed.shared/application/progresstilecallback.h (hunk 3 above)
Comment (achalpandeyy) [reply src/appleseed.shared/application/progresstilecallback.h]: Thanks for the review @dictoon!
I'll make the changes ASAP.

File: src/appleseed.studio/mainwindow/rendering/renderingmanager.h (hunk 5 above)
Comment (achalpandeyy) [reply src/appleseed.studio/mainwindow/rendering/renderingmanager.h]: If you are talking about header for `TileCallbackCollectionFactory`, I have already included it in `src/appleseed/renderer/api/rendering.h` and then added `rendering.h` in the file.

Do you still want me to include the header for `TileCallbackCollectionFactory` ?

And `#include <memory>` is already there for `std::unique_ptr`.

File: src/appleseed/renderer/kernel/rendering/tilecallbackcollection.h (hunk 8 above)
Comment (achalpandeyy) [reply src/appleseed/renderer/kernel/rendering/tilecallbackcollection.h]: Does that mean I am a part of appleseedhq? :joy:

File: src/appleseed.studio/mainwindow/rendering/renderingmanager.h (hunk 4 above)
Comment (achalpandeyy) [reply src/appleseed.studio/mainwindow/rendering/renderingmanager.h]: I don't know why this is happening. It seems perfect on my editor.

File: src/appleseed/renderer/kernel/rendering/tilecallbackcollection.h (hunk 8 above)
Comment (dictoon) [reply src/appleseed/renderer/kernel/rendering/tilecallbackcollection.h]: The comma between "YOUR NAME" and "The appleseedhq Organization" means "and" :)

But you're de facto part of "the appleseedhq contributors", yes.

File: src/appleseed.shared/application/progresstilecallback.h [hunk 10]
```diff
@@ -29,19 +29,16 @@
 
 #pragma once
 
+// appleseed.shared headers
```
Comment (dictoon) [inline src/appleseed.shared/application/progresstilecallback.h]: Move after appleseed.renderer headers. The order is: from most specific / highest level (i.e. project-specific files) to most general / lowest level (standard header files, platform header files).

File: src/appleseed.shared/application/progresstilecallback.cpp [hunk 11]
```diff
@@ -114,14 +115,14 @@ namespace
...
-                m_logger,
+                g_logger,
                 "rendering, %s done; about %s remaining...",
                 pretty_percent(m_rendered_pixels, total_pixels).c_str(),
                 pretty_time(remaining_time).c_str());
         }
 
       private:
-        Logger&                             m_logger;
+        Logger&                             g_logger;
```
Comment (dictoon) [inline src/appleseed.shared/application/progresstilecallback.cpp]: Use `m_` prefix for class members; `g_` is reserved for global variables.

File: src/appleseed.shared/application/progresstilecallback.cpp [hunk 12]
```diff
@@ -134,20 +135,28 @@ namespace
 // ProgressTileCallbackFactory class implementation.
 //
 
+struct ProgressTileCallbackFactory::Impl
+{
+    std::unique_ptr<renderer::ITileCallback> m_callback;
```
Comment (dictoon) [inline src/appleseed.shared/application/progresstilecallback.cpp]: `std::` qualifiers are redundant in this file since there's a `using namespace std` clause at the top.

Comment (dictoon) [reply src/appleseed.shared/application/progresstilecallback.cpp]: And so are `renderer::` qualifiers.

File: src/appleseed.studio/mainwindow/rendering/renderingmanager.h (hunk 5 above)
Comment (dictoon) [reply src/appleseed.studio/mainwindow/rendering/renderingmanager.h]: Oh sorry, my bad. In that case no we don't want to include `tilecallbackcollection.h` directly. And indeed <memory> is already included. Apologies!

File: src/appleseed/renderer/api/rendering.h [hunk 13]
```diff
@@ -45,4 +45,6 @@
 #include "renderer/kernel/rendering/nulltilecallback.h"
 #include "renderer/kernel/rendering/progressive/progressiveframerenderer.h"
 #include "renderer/kernel/rendering/tilecallbackbase.h"
+#include "renderer/kernel/rendering/tilecallbackcollection.h"
 #include "renderer/kernel/rendering/timedrenderercontroller.h"
+
```
Comment (dictoon) [inline src/appleseed/renderer/api/rendering.h]: Revert.

File: src/appleseed.studio/mainwindow/rendering/renderingmanager.h [hunk 14]
```diff
@@ -31,13 +31,13 @@
...
 // appleseed.studio headers.
 #include "mainwindow/rendering/qtrenderercontroller.h"
-#include "mainwindow/rendering/qttilecallback.h"
 #include "mainwindow/rendering/renderingtimer.h"
 
 // appleseed.renderer headers.
 #include "renderer/api/rendering.h"
 #include "renderer/api/types.h"
 #include "renderer/api/utility.h"
+#include "renderer/kernel/rendering/tilecallbackcollection.h"
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/rendering/renderingmanager.h]: Revert this (see my comment below).

File: src/appleseed/renderer/kernel/rendering/tilecallbackcollection.h (hunk 8 above)
Comment (achalpandeyy) [reply src/appleseed/renderer/kernel/rendering/tilecallbackcollection.h]: I'm honored. :)

File: src/appleseed.shared/application/progresstilecallback.cpp (hunk 12 above)
Comment (achalpandeyy) [reply src/appleseed.shared/application/progresstilecallback.cpp]: Oops, my bad! I didn't see those statements there.

**Comment by achalpandeyy** [comment]
> Almost there! The code is very clean, looking forward to merge this PR. Thanks!

Thanks for the review! I'll make the changes ASAP.

**Comment by achalpandeyy** [comment]
Thank you for the merge!
